build_docs.sh

```diff
@@ -0,0 +1,173 @@
+# folks that use build_docs.pl expect.
+DOCKER_RUN_ARGS+=('--user' "$(id -u):$(id -g)")
+
+# Running read-only with a proper tmp directory gives us a little
+# performance boost and it is simple enough to do.
+DOCKER_RUN_ARGS+=('--read-only')
+DOCKER_RUN_ARGS+=('--tmpfs' '/tmp')
+
+# Mount the docs build code so we can run it!
+DOCKER_RUN_ARGS+=('-v' "$DIR:/docs_build:cached")
+
+# Seccomp adds a *devestating* performance overhead if you happen
+# to have it installed.
+DOCKER_RUN_ARGS+=('--security-opt' 'seccomp=unconfined')
+
+# rewrite the arguments to be friendly to the docker container
+SAW_OUT=false
+NEW_ARGS=()
+RESOURCE_COUNT=0
+while [ $# -gt 0 ]; do
+  NEW_ARGS+=("$1")
+  case "$1" in
+  --doc)
+    shift
+    if [ ! -f "$1" ]; then
+      echo "Can't find $1"
+      exit 1
+    fi
+    DOC_FILE="$(to_absolute_path $1)"
+    GIT_REPO_ROOT="$(find_git_repo_root "$(dirname "$DOC_FILE")")"
+    DOCKER_RUN_ARGS+=('-v' "$GIT_REPO_ROOT:/doc:ro,cached")
+    NEW_ARGS+=("/doc${DOC_FILE/$GIT_REPO_ROOT/}")
+    ;;
+  --open)
+    DOCKER_RUN_ARGS+=('--publish' '8000:8000/tcp')
+    # Rituals to make nginx run on the readonly filesystem
+    DOCKER_RUN_ARGS+=('--tmpfs' '/run/nginx')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/log/nginx')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/body')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/fastcgi')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/proxy')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/uwsgi')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/scgi')
+    echo "------------------------ WARNING ------------------------"
+    echo "$(basename "$0") can't open a browser. It'll start the web" \
+         "server but you must open the browser yourself."
+    echo "------------------------ WARNING ------------------------"
+    ;;
+  --out)
+    shift
+    DOCKER_RUN_ARGS+=('-v' "$(dirname "$(to_absolute_path $1)"):/out:delegated")
+    NEW_ARGS+=("/out/$(basename "$1")")
+    SAW_OUT=true
+    ;;
+  --reference)
+    shift
+    if [ ! -d "$1" ]; then
+      echo "Can't find $1"
+      exit 1
+    fi
+    DOCKER_RUN_ARGS+=('-v' "$(to_absolute_path $1):/reference:ro,cached")
+    NEW_ARGS+=("/reference")
+    ;;
+  --rely_on_ssh_auth)
+    if [ "$SSH_AUTH_SOCK" != "" ]; then
+      # If we have SSH auth share it into the container.
+      if [ "$(uname -s)" != Linux* ]; then
+        echo "------------------------ WARNING ------------------------"
+        echo "Attempting to share ssh auth but this is unlikely to work" \
+             "outside of linux."
+        echo "------------------------ WARNING ------------------------"
+      fi
+      DOCKER_RUN_ARGS+=('-v' "$(dirname $SSH_AUTH_SOCK):$(dirname $SSH_AUTH_SOCK):ro")
+      DOCKER_RUN_ARGS+=('-e' "SSH_AUTH_SOCK=$SSH_AUTH_SOCK")
+    fi
+    # Mount our known_hosts file into the VM so it won't ask about github
+    DOCKER_RUN_ARGS+=('-v' "$(to_absolute_path ~/.ssh/known_hosts):/tmp/.ssh/known_hosts:ro,cached")
```

Comment:
I know we could use something like ` ssh-keyscan -t rsa github.com >> ~/.ssh/known_hosts` to add github's key but this feels, like, slightly better. I feel like this trusts that the user cloned from github and accepted the host. I could copy it in I suppose rather than mount. I've kind of gone mount crazy here but lots of mounts is kind of required for a project like this.